rfcs/006-token-based-extension-system.md

```diff
@@ -0,0 +1,44 @@
+# Token-based Extension System for Assessments
+
+## Introduction
+
+The purpose of this RFC is to propose a token-based extension system for assessments in order to provide students with a flexible mechanism for requesting deadline extensions. This system will allow students to utilize tokens they receive at the beginning of the semester to apply for extensions on their assessments. Instructors will have control over the allocation and usage of tokens within defined assessment sets, allowing them to manage and customize extension policies according to their course requirements.
+
+## Token Allocation
+
+2.1 Initial Allocation: At the start of the semester, all students will receive a fixed number of tokens. The exact number of tokens per student should be determined based on factors such as the course workload and assessment distribution.
+
+2.2 Additional Tokens: Instructors will have the discretion to assign additional tokens to individual students on a case-by-case basis if deemed necessary. This flexibility allows instructors to accommodate special circumstances or provide extra support to specific students.
+
+## Token Usage
+
+3.1 Extension Types: Each token will represent a specific type of extension, and the following options are proposed as initial suggestions:
+a) One token = one day extension
+b) One token = one week extension
+c) One token = student selects the new due date
+
+3.2 Token Constraints: Instructors will define the assessment sets where tokens can be utilized at the course instance level. Assessment sets can include various types of assessments such as homework, quizzes, exams, machine problems, etc. By specifying token constraints, instructors can control which assessments are eligible for extension requests using tokens. This provides instructors with the flexibility to tailor the extension policy to suit the nature and requirements of each assessment.
```

Comment:
Two questions here:

- I assume this is reusing PrairieLearn's existing notion of assessment sets? Or is this a separate token-specific notion of assessment sets? I _hope_ that assessment sets would be sufficient for instructors, but I do wonder if this would provide sufficient flexibility for students.
- You specifically mention quizzes and exams here. What would it mean for a student to use a token to extend a deadline for one of those? I'm imagining we wouldn't try to support extensions for assessments administered via a CBTF via PrairieTest. If so, we should call that out as an explicit non-goal.

Reply:
1) It will be reusing all the existing sets of assessments already present in PrarieLearn.
2) Yes, for the initial implementation we will only be working on the standard assessments such as group activities and pre-class quizzes etc. All the assessments administered through the CBTF via PrairieTest are not a part of the initial implementation.

Reply:
I think the tokens should be allowed to be applied to any PL set. If a instructor does not want it to be applied to a set that is called quizzes, they have the option to do that. I don't think we can pre-select the types of sets that can have this feature applied or not. 

Assessment that are controlled via PT will not appear on the assessment page outside the time the student is taking the assessment, so at least they won't be able to "request an extension" like the other assessments.